### PR TITLE
Synchronous Feedback Loop (breaking change)

### PR DESCRIPTION
**Pull request by andersio**
## TL;DR

1. Feedback Loops no longer require an asynchronous `ReactiveSwift.Scheduler`. Your unit test is impacted because of `Scheduler` no longer being required, but the behaviour should remain unchanged.

1. Unless you use the primitive `Feedback.init`, there is no change in API.

## Motivation
The primary motivation is to enable direct, hassle-free UI bindings with the feedback loop. At the moment, due to feedback events always being asynchronously scheduled, high volume of user interactions may outpace the feedback loop, causing glitches observable by users.

For example, let's assume a text view is fully integrated as part of the loop — _continuous_ changes hit the reducer, while `UITextView.text` is always updated with the text copy in the latest state. When the user types at a high pace, input events always take higher priority over dispatch queue callout (due to runloop callout order), to which the feedback events are scheduled. Therefore, this leads to the potential of the latest user input competing with previous versions that are awaiting on the main queue to be processed by the reducer.

### Example

In the best case scenario, RAF 0.7 will integrate fine with user inputs as one might expect:

Runloop Iteration | User Input | State | UITextView | Side Effect
---- | ---- | ---- | ---- | ----
t = 0 | - | abcde | abcde | -
t = 1 | "F" | abcde | abcdeF | `textViewDidChange(_:)` => enqueue "abcdeF"
 ⬆️ | - | abcdeF | abcdeF | consumed "abcdeF" & updated view
t = 2 | "G" | abcdeF | abcdeFG | `textViewDidChange(_:)` => enqueue "abcdeFG"
 ⬆️ | - | abcdeFG | abcdeFG | consumed "abcdeFG" & updated view
t = 3 | - | abcdeFG | abcdeFG | -

But if user types multiple keystrokes in the same runloop iteration, unpredictable outcomes may occur. One such possibility is as follows:

Runloop Iteration | User Input | State | UITextView | Side Effect
---- | ---- | ---- | ---- | ----
t = 0 | - | abcde | abcde | -
t = 1 | "F", "G", "H" | abcde | abcdeFGH | `textViewDidChange(_:)` => enqueue "abcdeF", "abcdeFG" and "abcdeFGH"
⬆️ | - | abcdeF | abcdeF | consumed "abcdeF" & updated text view
⬆️ | - | abcdeFG | abcdeFG | consumed "abcdeFG" & updated text view
⬆️ | - | abcdeFGH | abcdeFGH | consumed "abcdeFGH" & updated text view
t = 2 | "I", "J" | abcdeFGH | abcdeFGHIJ | `textViewDidChange(_:)` => enqueue "abcdeFI" and "abcdeFIJ"
⬆️ | - | abcdeFGHI | abcdeFGHI | consumed "abcdeFGHI" & updated text view
⬆️ | - | abcdeFGHIJ | abcdeFGHIJ | consumed "abcdeFGHIJ" & updated text view
t = 3 | - | abcdeFGHIJ | abcdeFGHIJ | -

As you may have observed, while the results are correct eventually, the copy in the text view is rolled back temporarily as the enqueued feedback events are drained & processed. This translates into glitches observed by users, where texts in the input field flicker.

### CODE REVIEW

**Comment by sergdort** [comment]
@andersio does it mean that we no longer need `Scheduler` for unit tests?

**Review by sergdort — APPROVED** [review]
tenor-129330753

👏 👏 👏

File: ReactiveFeedback/FeedbackEventCustomer.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+public class FeedbackEventConsumer<Event> {
```
Comment (AliSoftware) [inline ReactiveFeedback/FeedbackEventCustomer.swift]: [nit] file name is `Customer` instead of `Consumer` 😅

File: ReactiveFeedback/FeedbackEventCustomer.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    struct Token: Equatable {
+        let value: UUID
+
+        init() {
+            value = UUID()
+        }
+    }
+
+    func process(_ event: Event, for token: Token) {
+        fatalError()
```
Comment (AliSoftware) [inline ReactiveFeedback/FeedbackEventCustomer.swift]: [nit] would be nice to provide a message here
```suggestion
        fatalError("This is an abstract class. You must subclass this and provide your own implementation")
```

File: ReactiveFeedback/FeedbackEventCustomer.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    func unqueueAllEvents(for token: Token) {
+        fatalError()
```
Comment (AliSoftware) [inline ReactiveFeedback/FeedbackEventCustomer.swift]:
```suggestion
        fatalError("This is an abstract class. You must subclass this and provide your own implementation")
```

**Review by AliSoftware — COMMENTED** [review]
Nice!

**Comment by sergdort** [comment]
@andersio  Just realized we would need to observe on `UIScheduler` on all our ViewModels :s

**Review by sergdort — CHANGES_REQUESTED** [review]
Ok, I gave it another thought, what worries me now is that consumer would need to manage dispatching things on `MainQueue` themselves, where before when we were initialising the system with a scheduler and we were sure that we will receive updates on correct queue. 

I wonder if we can have optional or scheduler with default value that will make sure we observe things correctly?

**Review by sergdort — DISMISSED** [review]

**Comment by sergdort** [comment]
Hm actually we have `observe(on: UIScheduler())` in all our BentoVCs 👍

**Comment by sergdort** [comment]
So I've gave it a shot and integrated in our app, there is of course, gazillion tests failed because of assumptions that stuff is scheduled asynchronously. 

Also there are some UI issues because it now does emits multiple states synchronously mostly related to how we manage navigation as we usually have `intent` as a part of the state:

| Event | State |
|----| ----|
| ButtonPressed | state.intent = .shouldShowNextScreen |
| DidShowNextScreen (Triggered by the feedback which listens to the intent property) | state.intent = nil |

Because it emits state synchronously it causes to render screen multiple times with the same value of state which makes UI unresponsive.

So I'm proposing to add backwards compatibility by Introducing `FeedbackLoop.Feedback` so we can gradually transition to synchronous and deprecated old Feedback and Property initialiser?

What do you think @andersio @dmcrodrigues @mluisbrown @RuiAAPeres ?

**Comment by dmcrodrigues** [comment]
@sergdort seems a compelling argument, having a graceful period would definitely help with a gradual migration over one single and huge migration.

> Because it emits state synchronously it causes to render screen multiple times with the same value of state which makes UI unresponsive.

However, this part seems odd to me, if there's only two events, `buttonPressed` and `didShowNextScreen` how do we have so many "changes" of state? I would expect just 2 changes of state with will render exactly the same since the intent is not a UI state. I think would be good to understand why this is happening.

**Comment by mluisbrown** [comment]
I agree with @sergdort that we need to maintain some form of backwards compatibility, not just for making the transition easier for the Babylon app, but also for other users of RAF.

**Comment by andersio** [comment]
We can introduce a compatibility mode where existing behaviour is preserved if you supply a `Scheduler`. I think it should be doable with minimal change.

File: ReactiveFeedback/Feedback.swift [hunk 4]
```diff
@@ -28,15 +46,16 @@ public struct Feedback<State, Event> {
...
     ) where Effect.Value == Event, Effect.Error == Never {
-        self.events = { scheduler, state in
+        self.events = { state, scheduler, output in
             // NOTE: `observe(on:)` should be applied on the inner producers, so
             //       that cancellation due to state changes would be able to
             //       cancel outstanding events that have already been scheduled.
-            return transform(state)
-                .flatMap(.latest) { effects($0).producer.observe(on: scheduler) }
+            transform(state)
+                .flatMap(.latest) { effects($0).producer.observe(on: scheduler).enqueue(to: output) }
```
Comment (sergdort) [inline ReactiveFeedback/Feedback.swift]: Hey @andersio, I've added a scheduler parameter and default to `ImmediateScheduler` so we can retain previous behaviour when custom scheduler provided. What do you think?

Comment (andersio) [reply ReactiveFeedback/Feedback.swift]: ~LGTM~

Comment (andersio) [reply ReactiveFeedback/Feedback.swift]: I would recommend implementing it in `enqueue(to:)` though, with the scheduler supplied behind the public curtain. This way, if we are continuing to support queue confinement as an opt-in feature, the framework need not pass a scheduler to custom feedbacks, and custom feedbacks need not apply `observe(on: scheduler)` themselves (or very often forget to apply it until unexpected things happen).

**Comment by andersio** [comment]
I pushed the changes here, since I have no collaborator access.

https://github.com/andersio/ReactiveFeedback/commit/1c55434d7d45dd956651e581f38231910b37171b

File: ReactiveFeedback/FeedbackLoop.swift [hunk 5]
```diff
@@ -0,0 +1,182 @@
+import ReactiveSwift
+
+public final class FeedbackLoop<State, Event> {
```
Comment (sergdort) [inline ReactiveFeedback/FeedbackLoop.swift:3]: @andersio I've removed conformance to `PropertyProtocol` not sure why we need it?
Because it also gonna crash if you try to access the value before starting it.

Comment (andersio) [reply ReactiveFeedback/FeedbackLoop.swift:3]: oops

File: ReactiveFeedback/SignalProducer+System.swift [hunk 6]
```diff
@@ -15,16 +15,87 @@ extension SignalProducer where Error == Never {
...
+    /// Events are represented by an `Event` parameter. It represents all the possible Events that may occur in the System
+    /// Feedback defines an effect that may change the State.
+
+    /// - parameters:
+    ///     - initial: An initial `State` of the system.
+    ///     - scheduler: A Scheduler used for Events synchronisation
+    ///     - reduce: A function that produces a new State of a system by applying an Event
+    ///     - feedbacks: A Feedback loops that produces Events depending on the system's `State`
+    ///     - returns: A SignalProducer that emits current the state of the System
+    public static func feedbackloop<Event>(
```
Comment (andersio) [inline ReactiveFeedback/SignalProducer+System.swift]:
```suggestion
    public static func feedbackLoop<Event>(
```

File: ReactiveFeedback/SignalProducer+System.swift [hunk 7]
```diff
@@ -40,16 +111,11 @@ extension SignalProducer where Error == Never {
     ///     - reduce: A that produces a new State of the system by applying an Event
     ///     - feedbacks: A Feedback loops that produces Events depending on the system's state.
     ///     - returns: A SignalProducer that emits current the state of the System
-    public static func system<Event>(
+    public static func feedbackloop<Event>(
```
Comment (andersio) [inline ReactiveFeedback/SignalProducer+System.swift]:
```suggestion
    public static func feedbackLoop<Event>(
```

File: ReactiveFeedbackTests/SystemTests.swift [hunk 8]
```diff
@@ -7,10 +7,10 @@ class SystemTests: XCTestCase {
 
     func test_emits_initial() {
```
Comment (andersio) [inline ReactiveFeedbackTests/SystemTests.swift]: Should we restore the tests for the async version?

Comment (sergdort) [reply ReactiveFeedbackTests/SystemTests.swift]: Yeah 👍

**Comment by sergdort** [comment]
@andersio So it's ready to merge, but I was thinking if we should maybe change signature of the reducer to be `(inout State, Action) -> Void` from the perspective of mutability it's the same but it will become way easier for consumers as they won't need to do `var copy = state` dance or introduce `With`

What do you think @RuiAAPeres @dmcrodrigues @mluisbrown ?

**Comment by mluisbrown** [comment]
Absolutely in favour of changing reducer signature to be `(inout State, Action) -> Void`

**Comment by sergdort** [comment]
Because it's additive change it won't impact our codebase

**Comment by RuiAAPeres** [comment]
>  I was thinking if we should maybe change signature of the reducer to be (inout State, Action) -> Void

I am in favour of this. It makes its usage much more straightforward.

**Comment by DeFrenZ** [comment]
In my own state machine implementation I did switch to mutable state (in my case it was from `func update(for event: Event) -> (State, Action?)` to `mutable func update(for event: Event) -> Action?`) and it was a net QoL gain

**Comment by sergdort** [comment]
@DeFrenZ What is the difference between `Event` and `Action`?

**Comment by DeFrenZ** [comment]
In my case `Event` is the same as you intend (usually without a subtype like your `Action`), and my `Action` is basically a description of which feedback to perform. RAF injects feedbacks separately though so that part doesn't really match. My implementation was highly inspired by https://gist.github.com/andymatuschak/d5f0a8730ad601bcccae97e8398e25b2

**Comment by RuiAAPeres** [comment]
Approvals are for the weak. #ShipFastHandleTheConsequencesAtALaterPointInTime
